core/trino-spi/src/main/java/io/trino/spi/exchange/ExchangeSink.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi.exchange;
+
+import io.airlift.slice.Slice;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.concurrent.CompletableFuture;
+
+@ThreadSafe
+public interface ExchangeSink
+{
+    CompletableFuture<?> NOT_BLOCKED = CompletableFuture.completedFuture(null);
+
+    /**
+     * Returns a future that will be completed when the exchange sink becomes
+     * unblocked.  If the exchange sink is not blocked, this method should return
+     * {@code NOT_BLOCKED}
+     */
+    CompletableFuture<?> isBlocked();
```

Comment:
Per discussion of https://github.com/trinodb/trino/pull/10376#discussion_r776728449, do we want to make this parameterized on `partitionId`?

Reply:
I'm trying to understand if it is going to give us anything, or whether it will simply increase complexity.

Let's say there's a 2 way shuffle and we are about to write some data for `partition 0`. The sink is not ready to accept any data for `partition 0`, yet it is ready to accept some data for `partition 1`. Technically it is possible to temporarily stash the data for `partition 0` in some "buffer space" with a hope of finding some data for `partition 1`. However it is unclear how large that buffer space should be and whether it makes sense to add additional buffering on top of  what is implemented under the `ExchangeSink` interface. Thoughts?

Reply:
Does this mean we will return UNBLOCKED only when all partitions are ready to accept data?

Reply:
Yeah. We can change it later if we see we want it to be different.